### PR TITLE
Iteration2/board

### LINKED ISSUE #9: Board - Cells
As a board I need to have cells for the game to take place

### LINKED ISSUE #10: Board - Valid Coordinate?
As a board I want to know if a coordinate is valid so that I know if it can be used

### LINKED ISSUE #13: Board - Valid Placement?
As a board I want to tell if a ship's placement is valid so that it can be in play

A valid ship's placement must include ALL the following:
 - Same as ship length
 - Consecutive coordinates
 - Coordinates not diagonal
 - Overlapping ships

### LINKED ISSUE #15: Board - Place
As a board I want to be able to place a ship so that it has a location on the board

### LINKED ISSUE #16: Board - Render
As a board I want to be rendered so that I can relay information about the game to the player

### PR DESCRIPTION
**Pull request by AngelaGuardia**
Created all required methods and tests for the board.

The render method definitely needs some refactoring. Feel free to suggest changes!

### CODE REVIEW

File: lib/board.rb [hunk 1]
```diff
@@ -0,0 +1,92 @@
+class Board
+  attr_reader :cells, :height
+  def initialize(height = 4, width = 4)
+    @height = height
+    @width = width
+    @cells = generate_cells
+  end
+
+  def board_letters
+    ("A"..("A".ord + @height - 1).chr).to_a
```
Comment (ryandbarnett) [inline lib/board.rb:10]: This was confusing for me so...

("A"..("A".ord + @height - 1).chr).to_a
("A"..(65 + @height - 1).chr).to_a
("A"..(65 + 4 - 1).chr).to_a
("A"..(68).chr).to_a
("A".."D").to_a
["A","B","C","D"]

I didn't know we could do this with ord and chr. Interesting.

File: lib/board.rb [hunk 2]
```diff
@@ -0,0 +1,92 @@
...
+
+  def generate_cells
+    generated_cells = {}
+
+    board_letters.each do |letter|
+      board_numbers.each do |number|
+        generated_cells[letter + number] = Cell.new(letter + number)
+      end
+    end
+    generated_cells
```
Comment (ryandbarnett) [inline lib/board.rb]: Only thing I might change is use reduce instead of each on line 20.

Comment (ryandbarnett) [reply lib/board.rb]: I changed the generate_cells method to use reduce instead of each.

File: lib/board.rb [hunk 3]
```diff
@@ -0,0 +1,90 @@
...
+      end
+      generated_cells
+    end
+  end
+
+  def valid_placement?(ship, coordinates)
+    valid_coordinate?(coordinates) && valid_length?(ship, coordinates) && !overlapping_coordinates?(coordinates) && consecutive_coordinates?(ship, coordinates)
+  end
+
+  def valid_coordinate?(coordinates)
```
Comment (ryandbarnett) [inline lib/board.rb:30]: Would the name valid_coordinates? with an s be better since this method takes in an array of coordinates?

Comment (ryandbarnett) [reply lib/board.rb:30]: I looked at the valid_coordinate? method in the rubic and it expects to take a string of one coordinate not an array of multiple coordinates.

Comment (ryandbarnett) [reply lib/board.rb:30]: I can see why you need it to be able to accept multiple.

Comment (ryandbarnett) [reply lib/board.rb:30]: Maybe two methods
valid_coordinate?
and 
valid_coordinates?

Comment (AngelaGuardia) [reply lib/board.rb:30]: So looking back at this I realized that I had initially skipped over the valid_coordinate? portion in the rubric. When I wrote this method I thought I was super smart and had found an **additional** method I could write for robustness hahaha. 

I works for both single coordinates and multiple as long as they're stored in an array (single would be a one element array). I added tests for this.

The current name matched the rubric so I think we should keep it.

File: lib/board.rb (hunk 2 above)
Comment (AngelaGuardia) [reply lib/board.rb]: Agreed. Thanks for updating that.

**Comment by AngelaGuardia** [comment]
Really liked what you did with the render method refactoring!! Looks great. 

I'm adding an issue to write a test for the helper method that you added in your refactor so we can get to it later. Unless you feel like adding that right away haha.

**Pull request merged**